gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+
+	/*
+	 * We need to notify the GK blocks whenever we remove
+	 * a FIB entry that is accessible through a prefix.
+	 */
+	synchronize_gk_instances(ip_prefix_fib, gk_conf);
+
+	/*
+	 * From now on, GK blocks must not have a reference
+	 * to @ip_prefix_fib.
+	 */
+
+	switch (ip_prefix_fib->action) {
+	case GK_FWD_GRANTOR: {
+		ret = del_gateway_from_neigh_table_locked(
+			ip_prefix, GK_FWD_GATEWAY_BACK_NET,
+			ip_prefix_fib->u.grantor.eth_cache, gk_conf);
+		if (ret < 0)
+			return -1;
+
+		break;
+	}
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET: {
+		ret = del_gateway_from_neigh_table_locked(
+			ip_prefix, ip_prefix_fib->action,
+			ip_prefix_fib->u.gateway.eth_cache, gk_conf);
+		if (ret < 0)
+			return -1;
+
+		break;
+	}
+
+	case GK_DROP:
+		break;
+
+	/*
+	 * For GK_FWD_NEIGHBOR_*_NET FIB entries, they are initialized
+	 * when the Gatekeeper starts. These FIB entries are only reserved
+	 * for the network prefixes, for which the Gatekeeper is responsible.
+	 * If we change the network prefixes, then the Gatekeeper may
+	 * need to restart, so one can ignore deletion of these FIB entries.
+	 */
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		return -1;
+		break;
+	}
+
+	/* Reset the fields of the deleted FIB entry. */
+	initialize_fib_entry(ip_prefix_fib);
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: delete a fib entry [ip prefix = %s]\n", ip_prefix->str);
+
+	return 0;
+}
+
+/*
+ * Fill up the Ethernet cached header.
+ * Note that the destination MAC address should be filled up by LLS.
+ */
+static inline void
+fill_up_ether_cache_locked(struct ether_cache *eth_cache,
+	struct ipaddr *addr, struct gatekeeper_if *iface)
+{
+	eth_cache->filled = true;
+	eth_cache->stale = true;
+	rte_memcpy(&eth_cache->ip_addr, addr, sizeof(eth_cache->ip_addr));
+	eth_cache->eth_hdr.ether_type = addr->proto;
+	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+	eth_cache->ref_cnt = 1;
+}
+
+/* Create a FIB entry for the gateway. */
+static struct gk_fib *
+create_fib_entry_for_gateway_locked(struct ipaddr *ip_addr, int prefix_len,
+	enum gk_fib_action action, struct ipaddr *gw_addr,
+	struct gatekeeper_if *iface, struct gk_config *gk_conf)
+{
+	int ret, fib_id;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gk_fib *gw_fib;
+	struct ether_cache *eth_cache;
+	struct neighbor_hash_table *neigh_ht;
+
+	/* Find the neighbor FIB entry for this gateway. */
+	struct gk_fib *neigh_fib = find_fib_entry_for_neighbor_locked(
+		gw_addr, action, iface, gk_conf);
+	if (neigh_fib == NULL)
+		return NULL;
+
+	/* Find the Ethernet cached header entry for this gateway. */
+	neigh_ht = &neigh_fib->u.neigh;
+	eth_cache = neigh_get_ether_cache(neigh_ht, &gw_addr->ip);
+	if (eth_cache == NULL)
+		return NULL;
+
+	/* Find an empty FIB entry for the Gateway. */
+	fib_id = get_empty_fib_id(ip_addr->proto, gk_conf);
+	if (fib_id < 0)
+		return NULL;
```

Comment:
If `neigh_get_ether_cache()` added an entry, you need to release it before returning an error.